src/main/java/com/uid2/admin/vertx/service/ServiceLinkService.java

```diff
@@ -118,6 +128,103 @@ private void handleServiceLinkAdd(RoutingContext rc) {
+            }
+
+            if (siteProvider.getSite(siteId) == null) {
+                ResponseUtil.error(rc, 404, "site_id " + siteId + " not valid");
+                return;
+            }
+
+            final List<ServiceLink> serviceLinks = this.serviceLinkProvider.getAllServiceLinks()
+                    .stream().sorted(Comparator.comparing(ServiceLink::getLinkId))
+                    .collect(Collectors.toList());
+
+            ServiceLink serviceLink = serviceLinks
+                    .stream().filter(s -> s.getServiceId() == serviceId && s.getSiteId() == siteId && s.getName().equals(name))
+                    .findFirst()
+                    .orElse(null);
+            if (serviceLink == null) {
+                ResponseUtil.error(rc, 404, "failed to find a service_link for serviceId: " + serviceId + ", site_id: " + siteId + " and name: " + name);
+                return;
+            }
+
+            if (serviceLinks.stream().anyMatch(sl -> sl.getServiceId() == serviceId && sl.getLinkId().equals(linkId))) {
+                ResponseUtil.error(rc, 400, "service link id already exists for service_id: " + serviceId);
+                return;
+            }
+
+            if (linkId != null && !linkId.isEmpty()) {
+                serviceLink.setLinkId(linkId);
+            }
+
+            storeWriter.upload(serviceLinks, null);
+
+            rc.response().end(toJson(serviceLink).encodePrettily());
+        } catch (Exception e) {
+            ResponseUtil.errorInternal(rc, "Internal Server Error", e);
+        }
+    }
+
+    private void handleServiceLinkDelete(RoutingContext rc) {
+        try {
+            serviceProvider.loadContent();
+            JsonObject body = rc.body().asJsonObject();
+            if (body == null) {
+                ResponseUtil.error(rc, 400, "json payload required but not provided");
+                return;
+            }
+
+            String linkId = body.getString("link_id");
+            Integer serviceId = body.getInteger("service_id");
+
+            if (linkId == null || serviceId == null) {
+                ResponseUtil.error(rc, 400, "required parameters: link_id, service_id");
+                return;
+            }
+
+            ServiceLink serviceLink = this.serviceLinkProvider.getAllServiceLinks()
+                    .stream().filter(s -> s.getServiceId() == serviceId && s.getLinkId().equals(linkId))
+                    .findFirst()
+                    .orElse(null);
+            if (serviceLink == null) {
+                ResponseUtil.error(rc, 404, "failed to find a service_link for serviceId: " + serviceId + " and link_id: " + linkId);
+                return;
+            }
+
+            final List<ServiceLink> serviceLinks = this.serviceLinkProvider.getAllServiceLinks()
+                    .stream().sorted(Comparator.comparingInt(ServiceLink::getServiceId))
+                    .collect(Collectors.toList());
+
+            serviceLinks.remove(serviceLink);
```

Comment:
do we want to keep track of which services  and links were created and deleted in the past ? If that is useful then we should consider soft deletion / disabling them

Reply:
I'm not sure there is currently a need to keep track of disabled keys. A number of other entities also support deleting the key

Reply:
Ok, I feel a little uneasy about not leaving a trace of the deleted key and who deleted it. Could we at least write a log message to record this info ?

Reply:
I don't disagree, but there is no auditing anywhere in Admin, so I think this is a bigger conversation. We also want to move away from using Github as the authentication mechanism.